### DISCUSSION TITLE
Thoughts on TvOS and WatchOS support

### DISCUSSION THREAD

**Discussion by Schlaubischlump** [BLeeEZ/amperfy#303] [Ideas]
Are there any plans for a TvOS or WatchOS version?

I would gladly help with any development for these platforms. However, I never personally developed an TvOS App and my watchOs experiences are also pretty limited. That being said, I do have plenty of experience with AppKit and UlKit, so it shouldn't be too hard to pick up these platforms. 
Based on my understanding, TVOS is probably the easier target since it supports a subset of UlKit. At least AmperfyKit should be easy to port and parts of the UI might also carry over.
For watchOS however, things are more difficult. It probably requires a full UI rewrite in SwiftUI. Maybe a stripped down version of Amperfy that is configured using the iOS app would be a good starting point? Also AmperfyKit uses UIKit in some places. While UIImage is available and there is WKInterfaceDevice to replace UIDevice, UIView for example, is not available. So that might be problematic as well.

Any thoughts on that? E.g. Is it feasible to extract all UI specific code (aka code that uses UIView) from AmperfyKit?

**Comment by BLeeEZ** [comment]
In principle I am open for all platforms. I personally don't own an AppleTV nor AppleWatch. This makes testing and further bug fixes difficult for me. AmperfyKit can be adjusted to get work on those platforms. I think we can fix that UIView issue.
One other big problem are "App Groups". I try to avoid "App Groups" due to management in App Connect. This limitation makes it difficult to exchange data with watchOS. TVOS should not be affected by that.

**Reply by fivestones** [reply]
I don't have any firsthand experience, but from what ChatGPT tells me you can skip using App Groups entirely if you are using WCSession to do file transfers between iOS and watchOS.

**Comment by Schlaubischlump** [comment]
I tried to compile amperfy for TvOS but it is next to impossible. To many things are unavailable. The code becomes an unreadable mess with all the #if and I'm not even close to getting it to compile. 

Maybe it is easier to add a subfolder or project for TvOS, write specific UI code for the platform and only add a single #if to the scene delegate. 

That means we would effectively end up with two separate code bases that might share common code where possible, but would be mostly separate. 

What do you think?

**Comment by BLeeEZ** [comment]
Thank you for the investigation. 
I don’t have an Apple TV. I don’t like to publish code on a device that I can’t test for.
I encountered enough bugs that occurred only on real devices, which makes simulator only not feasible.
So from my POV I would postpone Apple TV support.

**Comment by Schlaubischlump** [comment]
I do understand not wanting to publish for a platform you cannot debug on. There are always device specific bugs the simulator doesn't catch. 

Nevertheless, I will probably move forward with implementing a TvOS version, since I need it for personal use. I'm not sure if I have enough time to finish it all by myself, but time will tell 😄

Wether you want to merge these efforts in the end is up to you. I can totally understand not wanting to maintain a platform you have no access to.

**Comment by Schlaubischlump** [comment]
There is now an app called SubSwift in the TvOS AppStore. It costs 5,99€. It implements the SubSonic API and works nicely with NextCloud music.

**Reply by ciureaalex** [reply]
Yeah but currently it has a few issues in regards of its playback capability (+ some UX issues).
In particular:
- to enable AirPlay as a destination you must first connect over Bluetooth if no "TV speakers" are available and then AirPlay playback will work up to a point where everything bugs out;
- audio seems to not be clear even in "RAW" mode. (Although that toggle seems broken?)

Still better than nothing but I do hope either Amperfy or SubSwift or any other app gets to work as well on tvOS as they work on macOS or iOS.

I already contacted SubSwift's developer regarding the AirPlay issue with my findings and a mention on what other apps do it with no issues since at least one of them has its source already available on GitHub. (Assuming it would help the dev)

**Comment by jacobdanielrose** [comment]
I'm still in the learning phase of SwiftUI and developing for apple in general but I would be very much interested in adding WatchOS support and I have an Apple Watch so I can debug on device.

**Reply by TheMBeat** [reply]
Any progress?

**Reply by jacobdanielrose** [reply]
No, life kinda got away from me at the moment. If I find the time to get around to it at some point, I'll update this thread tho.
